### ISSUE TITLE
[29/05/2020] Friday 9pm GMT+2 ELECTRA: Pre-training Text Encoders as Discriminators Rather Than Generators

### ISSUE THREAD

**Issue by Omarito2412** [eg-nlp-community/nlp-reading-group#14]
Please join us in our discussion of "ELECTRA: Pre-training Text Encoders as Discriminators Rather Than Generators" on Friday at 9pm GMT+2.

URL: https://openreview.net/forum?id=r1xMH1BtvB

**Comment by DiaaMohsen** [comment]
I thought last reading group was an exception to be held on Friday so are
we going to make it on Friday for future ones?

**Comment by Omarito2412** [comment]
@DiaaMohsen Nope, let's just have it on Friday this week and settle for a new slot with a vote

**Comment by maimagdi** [comment]
Thank you guys you mentioned a paper with a similar "concept" that used tf-idf could you share it please? [I see the ELECTRA point of you, just want to learn] 
and can you share the competition link you mentioned? :D 
(not sure if I'm posting in the right place)

**Comment by ibrahimsharaf** [comment]
- New language model training objective is proposed => Replaced (fake) token(s) prediction.
- A dual-network architecture, the first one (generator) which is a small transformer (BERT) model, for typical masked language modelling, the second (discriminator), applies binary classification on the sampled text from the generator (predicts whether each token is original or replaced).
- For fine-tuning, only the discriminator (the 2nd network) is used, no mask tokens (solves BERT's pretraining/fine-tuning mask token discrepancy issue).
- Same number of params as (BERT, XLNET, roBERTa...), more computationally efficient (25% of the training time) => SOTA scores on GLUE and SQUAD (1.1, 2.0).

**Comment by hadyelsahar** [comment]
tldr; Electra works for those two reasons:
- the discriminator learns from all the tokens not only the masked 15% at each sample
- Discriminator has never seen those weird special tokens such as [Mask] and [sep] so there's no pretraining / fine tuning discrepancy 

A good question has been proposed about Why not using an even simpler generator like a unigram language model or filling the gap using Mutual information with the context: 
A good generator needed but not a very strong one that still makes mistakes so the Discriminator can still learn something this is explained in this figure:
 
![image](https://user-images.githubusercontent.com/1453243/83302298-5994ab00-a1fb-11ea-84d9-01fb539a90b9.png)

**My take on using REINFORCE for Adversarial training.** 

REINFORCE is a reward maximization algorithm, training a GAN generator with it will tend to converge to a small set of high quality low diversity samples, this phenomena has been explained in detail in [Language GANs falling short paper](https://arxiv.org/abs/1811.02549) (although in the Masked language models is a more constrained setup that might enforce more context diversity). 

To overcome this problem some alternatives (although little in language) has been explored such as tricks for MCMC sampling and [Distributional policy gradients](https://arxiv.org/pdf/1912.08517.pdf)

**Comment by hadyelsahar** [comment]
**Introduction to Energy-Based Models (EBMs)**

why: 
![image](https://user-images.githubusercontent.com/1453243/83341942-5e378d00-a2e9-11ea-8672-b3a4dc3f5a45.png)

what: 
2006 Tutorial on EBMs: http://yann.lecun.com/exdb/publis/pdf/lecun-06.pdf
A Recent Yan le cun Talk: https://www.youtube.com/watch?v=A7AnCvYDQrU
